File: backend/app/schemas/train.py
from pydantic import BaseModel


class TrainNode(BaseModel):
	id: str
	name: str
	type: str  # 'IC' | 'REGIONAL' | 'FREIGHT'
	originStationId: str
	destinationStationId: str
	direction: str  # 'outbound' | 'return'
	currentStationId: str
	nextStationId: str | None = None
	currentSegmentId: str | None = None
	progress: float
	status: str  # 'dwelling' | 'running' | 'waiting' | 'derailed'
	routeStationIds: list[str] = []
	routeSegmentIds: list[str] = []
	routeIndex: int = 0
	vmax: int
	priority: int
	massTonnes: float
	lengthM: int
	accel: float
	decel: float
	dwellUntil: float | None = None
	delayedByEventId: str | None = None
	updatedAt: float


class TrainsSnapshotResponse(BaseModel):
	trains: list[TrainNode]
	timestamp: float

File: frontend/src/lib/types/train.ts
export type TrainType = 'IC' | 'REGIONAL' | 'FREIGHT';
export type TrainDirection = 'outbound' | 'return';
export type TrainStatus = 'dwelling' | 'running' | 'waiting' | 'derailed';

export type TrainNode = {
	id: string;
	name: string;
	type: TrainType;
	originStationId: string;
	destinationStationId: string;
	direction: TrainDirection;
	currentStationId: string;
	nextStationId: string | null;
	currentSegmentId: string | null;
	progress: number;
	status: TrainStatus;
	routeStationIds: string[];
	routeSegmentIds: string[];
	routeIndex: number;
	vmax: number;
	priority: number;
	massTonnes: number;
	lengthM: number;
	accel: number;
	decel: number;
	dwellUntil: number | null;
	delayedByEventId: string | null;
	updatedAt: number;
};

export type TrainsSnapshotResponse = {
	trains: TrainNode[];
	timestamp: number;
};
